rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,650 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use std::sync::Arc;
+
+use arrow::array::ArrayData;
+use arrow::datatypes::{ArrowPrimitiveType, UInt16Type, UInt32Type, UInt64Type, UInt8Type};
+use arrow::util::bit_util::ceil;
+use arrow_array::{cast::AsArray, Array, ArrayRef, PrimitiveArray};
+use arrow_schema::DataType;
+use bytes::{Bytes, BytesMut};
+use futures::future::{BoxFuture, FutureExt};
+use log::trace;
+use num_traits::{AsPrimitive, PrimInt};
+use snafu::{location, Location};
+
+use lance_arrow::DataTypeExt;
+use lance_core::{Error, Result};
+
+use crate::encoder::EncodedBufferMeta;
+use crate::{
+    decoder::{PageScheduler, PrimitivePageDecoder},
+    encoder::{BufferEncoder, EncodedBuffer},
+};
+
+// Compute the number of bits to use for each item, if this array can be encoded using
+// bitpacking encoding. Returns `None` if the type or array data is not supported.
+pub fn num_compressed_bits(arr: ArrayRef) -> Option<u64> {
+    match arr.data_type() {
+        DataType::UInt8 => Some(num_bits_for_type::<UInt8Type>(arr.as_primitive())),
+        DataType::UInt16 => Some(num_bits_for_type::<UInt16Type>(arr.as_primitive())),
+        DataType::UInt32 => Some(num_bits_for_type::<UInt32Type>(arr.as_primitive())),
+        DataType::UInt64 => Some(num_bits_for_type::<UInt64Type>(arr.as_primitive())),
+        // TODO -- eventually we could support signed types as well
+        _ => None,
+    }
+}
+
+// Compute the number of bits to to use for bitpacking generically.
+fn num_bits_for_type<T>(arr: &PrimitiveArray<T>) -> u64
+where
+    T: ArrowPrimitiveType,
+    T::Native: PrimInt + AsPrimitive<u64>,
+{
+    // safe to unwrap here because if the array is empty or all nulls it's
+    // handled in BasicEncoder
+    let max = arrow::compute::bit_or(arr).unwrap();
+    let num_bits = arr.data_type().byte_width() as u64 * 8 - max.leading_zeros() as u64;
+
+    // we can't bitpack into 0 bits, so the minimum is 1
+    num_bits.max(1)
+}
+
+#[derive(Debug)]
+pub struct BitpackingBufferEncoder {
+    num_bits: u64,
+}
+
+impl BitpackingBufferEncoder {
+    pub fn new(num_bits: u64) -> Self {
+        Self { num_bits }
+    }
+}
+
+impl BufferEncoder for BitpackingBufferEncoder {
+    fn encode(&self, arrays: &[ArrayRef]) -> Result<(EncodedBuffer, EncodedBufferMeta)> {
+        // calculate the total number of bytes we need to allocate for the destination.
+        // this will be the number of items in the source array times the number of bits.
+        // let count_items = count_items_to_pack(arrays);
+        let count_items = arrays.iter().map(|arr| arr.len()).sum::<usize>();
+        let dst_bytes_total = count_items * ceil(self.num_bits as usize, 8);
```

Comment:
Would `ceil(count_items * self.num_bits as size, 8)` be more precise?  In other words, if you have 10 items and each requires 3 bits you would want 4 bytes (32 bits) and not 10 bytes?

Reply:
Yeah you're right.. good call, made this change